[WIP] Refactor scaler code

untom opened the pull request:
This PR refactors the scalers in `preprocessing` to use a common baseclass (`BaseScaler`). This is done to share some common functionality / avoid code duplication. This PR is part of a series of PR that split up #2514 , and other PRs in this series intend to introduce new scalers that will make use of `BaseScaler` (e.g. a scaler that uses robust statistics).

There is one issues with this PR that needs to be discussed:

All scalers (both the existing ones and the one's I'm going to introduce) do center the data and then scale it to fall into some range. So in essence, all layers do

```
(x - some_centering_statistic) / some_scaling_statistic
```

for each feature-column x.  The existing scalers already have attributes that expose their centering/scaling statistic, however those are not uniformly named.  I propose that the  `StandardScaler.mean_`/`StandardScaler.std_`/`MinMaxScaler.scale_` and `MinMaxScaler.min_` properties are being deprecated and replaced with `*.center_`/`*.scale_`, and the `with_mean`/`with_std` constructer arguments be renamed to `with_centering`/`with_scaling`.

@larsmans already said back in #2514 that he is against renaming the `mean_`/`std_` attributes, since the scaler API should be stable. An additional problem is that my proposal would change the meaning of `MinMaxScaler.scale_` (as the previous implementation used a slightly different math to arrive at the same scaling result).   This will of course break user code that uses `MinMaxScaler.scale_` and relies on its current implementation.  (NOTE: I had overlooked this when proposing #2514 and just noticed it when preparing this PR)

There are several ways to deal with this:
- don't expose the new `scale_`/`center_` arguments in StandardScaler/MinMaxScaler, and just keep the old `mean_`/`std_`/`scale_` arguments.
- deprecate `mean_`/`std_` and remove them after a few releases. Print a warning   whenever a user uses `scale_` on `MinMaxScaler` that the value has changed.
- ???

I would really appreciate any input on what the right thing to do here is!

_NOTE:_ As side-effect of this PR, `StandardScaler` and `MinMaxScaler` gain the ability to scale on other rows/samples instead of just columns/features (i.e., it is possible to use `axis=1`). However, no tests for this new functionality is included in this PR. This is because I intend to send another PR tomorrow that refactors the tests in this module (and adds the tests for `axis = 1`). I just thought that splitting up the tests makes it easier to review the changes (but I can add that other PR to this one if you wish).

Commit:
ENH Refactor scaler code

amueller commented:
So the idea of this PR is to do two things, right?

1. Make the attributes of the scalers consistent
2. refactor their code.

Shouldn't that result in less code, not more?

untom commented:
@amueller: Yes, although this PR is mainly meant to pave the way to adding more scalers (Concretely, `MaxAbsScaler` and `RobustScaler`). While the refactorings IMHO already make the code simpler, they are only going to really pay off once those are added to the codebase.

As far as "making the attributes consistent" goes, the consensus seems to be that it's too late to change this, so I undid those changes.


FYI, I don't know why it says "This pull request contains merge conflicts that must be resolved. "... at least on my local branch I can cleanly merge this into master.

untom commented:
>>   Or remove it completely, and have the xxx_scale function transpose their arguments if needed.
>
> +1
>

@ogrisel: I've changed the code to do it this way.

jnothman commented:
> FYI, I don't know why it says "This pull request contains merge conflicts that must be resolved. "... at least on my local branch I can cleanly merge this into master.

Do a `git rebase master` then `git push -f`

untom commented:
I tried that, but it doesn't seem to work:

    untom@comp:scikit-learn$ git rebase master
    Current branch refactor_scaling is up to date.
    untom@comp:scikit-learn$ git push -f
    Username for 'https://github.com': untom
    Password for 'https://untom@github.com': 
    Counting objects: 8, done.
    Delta compression using up to 8 threads.
    Compressing objects: 100% (8/8), done.
    Writing objects: 100% (8/8), 2.25 KiB | 0 bytes/s, done.
    Total 8 (delta 7), reused 0 (delta 0)
    To https://github.com/untom/scikit-learn.git
     + 4c05f6b...6570389 refactor_scaling -> refactor_scaling (forced update)

jnothman commented:
Your master may not be up-to-date:

In case you need it, first `git remote add upstream https://github.com/scikit-learn/scikit-learn`

Then
```
git pull upstream master:master
git checkout refactor_scaling
git rebase master
git push -f
```

untom commented:
Thanks, this did the trick! Not sure why it didn't work before (could've sworn I've pulled from upstream before my first try).

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -171,18 +187,21 @@ class MinMaxScaler(BaseEstimator, TransformerMixin):
         self.copy = copy
 
-    def fit(self, X, y=None):
+    def fit(self, X, y=None, copy=None):
```
Why did you add this fit parameter?

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -171,18 +187,21 @@ class MinMaxScaler(BaseEstimator, TransformerMixin):
 
     Attributes
     ----------
-    min_ : ndarray, shape (n_features,)
```
All attributes should be documented or deprecated.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -191,57 +210,33 @@ def fit(self, X, y=None):
+        self.scalefactor_ = data_range / (feature_range[1] - feature_range[0])
+        self.center_ = data_min - feature_range[0] * self.scalefactor_
+
         self.scale_ = (feature_range[1] - feature_range[0]) / data_range
```
If these are here but not used, I guess they should be read-only attributes maybe to make explicit they they are not used?

untom replied:
I'm not sure what you are referring to. Could you elaborate?

amueller replied:
``scale_`` and ``min_`` are attributes that are set but not used in ``transform``, right?
So an unsuspecting user might try to change ``min_`` to change the behavior of the estimator, but nothing will happen. Also I'm wondering whether these names should be deprecated then.

untom replied:
> Also I'm wondering whether these names should be deprecated then.

My original proposal did deprecate these names, but I think the consensus was that it is too late to fundamentally change the existing API.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -39,116 +45,122 @@
+            X = X.astype(np.float)
+        if sparse.issparse(X):
+            if not (sparse.isspmatrix_csc(X) or sparse.isspmatrix_csr(X)):
+                raise TypeError("Scaling only supports CSC and CSR "
```
This line can not be reached, right?

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -299,101 +294,49 @@ class StandardScaler(BaseEstimator, TransformerMixin):
+            used for later scaling along the specified axis.
         """
-        check_is_fitted(self, 'std_')
+        self.center_ = None
```
why is this initialization necessary?

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -502,6 +445,59 @@ def transform(self, X, y=None):
+    Instead the caller is expected to either set explicitly
+    `with_centering=False` (in that case, only variance scaling will be
+    performed on the features of the CSR matrix) or to call `X.toarray()`
+    if he/she expects the materialized dense array to fit in memory.
```
I would use passive or "they"

untom committed:
Make some MinMaxScaler attributes read-only

amueller commented:
Sorry for the lack of feedback. What is the status here? I lost track a bit.
I want to add a ``MaxAbsScaler`` so we can scale sparse data, and I'm wondering if I should wait for this refactoring.

amueller commented:
So there is a ``MaxAbsScaler`` in #2514 ... hum... Are you still working on this?

untom commented:
Ok, quick overview of this: My original PR for this was #2514, which added two new scalers  (`RobustScaler` and `MaxAbsScaler`) and refactored the code to avoid code duplication (plus a few other niceties for the scalers). However according to reviewers that was too big of a change to get it accepted/merged in one bit, so I split it up into several PRs. This one (#3639) was meant to be the first one: it refactors the existing scalers to make it easier to add new ones. However, I never heard back from any reviewers after all of their initial concerns had been addressed.

Thus I decided to submit new scalers first, figuring the code could always be refactored later (when the usefulness of the refactoring became more apparent).  In #4125 I tried submitting the `RobustScaler` without making larger changes to the internal code layout. When that PR also pretty much died the same way as #3639 , I tried pinging reviewers and fixed merge-conflicts with the master branch as they came up. But after a while I pretty much gave up on the whole thing, and thus never submitted a separate `MaxAbsScaler` PR.

At this point I am quite frustrated with the whole process, as I have wasted a lot of time on this throughout the 1.5 years since the first submission of the code. I don't know what more I should have done to get any of these PR merged, but I'd be happy for any feedback. If there's anything I can do to make this easier I'd love to hear it for future PRs.

With that said, If there is true interest in `MaxAbsScaler`, I can send it as a separate PR this weekend.  I still think that MaxAbsScaler and RobustScalers would be nice additions to sklearn and I'm happy to put in the work, but I'd like to avoid wasting any more time in these PRs if they are going to slowly die due to lack of reviewer-interest, like they have in the past.

amueller commented:
I am very sorry for the lack of feedback, and I totally understand your frustration. We are really starved for devs / reviewers at the moment. I think bost RobustScaler and MaxAbsScaler would be great additions.
I am afraid that if you open another PR, it might not get the attention it deserves, and create even more frustration. Lets get #4125 merged first, and then go for the MaxAbsScaler? I just saw that you pinged me a month ago. Sorry for my lack of responsiveness.

untom commented:
Sounds good! Thanks for taking the time to review #4125 :)

jnothman commented:
I must say I am also a bit frustrated after going through all that work on #2514 with you. Hence I'm very glad to see #4125 finally merged. Congrats!

amueller commented:
Now lets get #4828 in next :)

untom commented:
I think most (all?) of the patches are in by now, so I'm going to close this PR.

amueller commented:
well you also introduced a base-class that got rid of some code duplication....

amueller commented:
Thank you for your patients and all your work btw :)

untom commented:
True, but introducing that base-class would've required renaming the `mean_` and `std_` fields of `StandardScaler` to `center_` and `scale_` and changed the meaning of the `scale_` field of `MinMaxScaler`, and there was quite some pushback to that. While those restrictions could've been worked around, I felt that it was not worth it.

amueller commented:
fair enough :) thank again!
